vmm/sandbox/src/cgroup.rs

```diff
@@ -0,0 +1,425 @@
+        let mut sandbox_cgroup_path = Path::new(sandbox_cgroup_path_str.as_str());
+        // CgroupBuilder::new() func doesn't accept the cgroup name has "/" prefix,
+        // So need to remove the "/" prefix for sandbox_cgroup_path
+        if sandbox_cgroup_path.has_root() {
+            sandbox_cgroup_path = sandbox_cgroup_path.strip_prefix("/")?;
+        }
+
+        let sandbox_cgroup =
+            CgroupBuilder::new(sandbox_cgroup_path.to_str().unwrap()).build(all_hiers)?;
+
+        // Only create the vcpu and pod_overhead cgroups in the cpu cgroup subsystem
+        let vcpu_cgroup_path = format!(
+            "{}/{}",
+            sandbox_cgroup_path.to_str().unwrap(),
+            VCPU_CGROUP_NAME
+        );
+        let vcpu_cgroup =
+            CgroupBuilder::new(vcpu_cgroup_path.as_str()).build(cpu_subsystem_hier())?;
+
+        let pod_overhead_cgroup_path = format!(
+            "{}/{}",
+            sandbox_cgroup_path.to_str().unwrap(),
+            POD_OVERHEAD_CGROUP_NAME
+        );
+        let pod_overhead_cgroup =
+            CgroupBuilder::new(pod_overhead_cgroup_path.as_str()).build(cpu_subsystem_hier())?;
+
+        Ok(SandboxCgroup {
+            sandbox_cgroup,
+            vcpu_cgroup,
+            pod_overhead_cgroup,
+        })
+    }
+
+    pub fn update_res_for_sandbox_cgroups(&self, sandbox_data: &SandboxData) -> Result<()> {
+        // apply the total resources = sum(sum(containers_resources) + pod_overhead)) in the sandbox cgroup dir
+        if let Some(total_resources) = get_total_resources(sandbox_data) {
+            apply_cpu_resource(&self.sandbox_cgroup, &total_resources)?;
+            apply_memory_resource(&self.sandbox_cgroup, &total_resources)?;
+            apply_cpuset_resources(&self.sandbox_cgroup, &total_resources)?;
+            apply_hugetlb_resources(&self.sandbox_cgroup, &total_resources)?;
+        }
+
+        // apply the cpu resource of containers in the vcpu cpu subsystem cgroup
+        if let Some(containers_resources) = get_resources(sandbox_data) {
+            apply_cpu_resource(&self.vcpu_cgroup, containers_resources)?;
+        }
+
+        // apply the cpu resource of pod_overhead in the pod_overhead cpu subsystem cgroup
+        if let Some(overhead_resources) = get_overhead_resources(sandbox_data) {
+            apply_cpu_resource(&self.pod_overhead_cgroup, overhead_resources)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn add_process_into_sandbox_cgroups(
+        &self,
+        pid: u32,
+        vcpu_threads: VcpuThreads,
+    ) -> Result<()> {
+        // Add vmm process into the sandbox_cgroup
+        self.sandbox_cgroup.add_task_by_tgid((pid as u64).into())?;
+
+        // Move vmm process from parent sandbox cgroup into pod_overhead cgroup
+        // Then move the all vcpu threads of vmm process into vcpu cgroup
+        self.pod_overhead_cgroup
+            .add_task_by_tgid((pid as u64).into())?;
```

Comment:
Is `sandbox_cgroup.add_task_by_tgid()` necessary after calling `pod_overhead_cgroup.add_task_by_tgid()`

Reply:
Yes, `sandbox_cgroup.add_task_by_tgid()` is applied for all hier in the `sandbox_cgroup` controller, while `pod_overhead_cgroup.add_task_by_tgid()` is explicitly processed for `CPU subsytem`.